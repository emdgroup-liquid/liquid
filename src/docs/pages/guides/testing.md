---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

