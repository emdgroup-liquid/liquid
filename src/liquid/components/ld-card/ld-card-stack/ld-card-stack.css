:host,
.ld-card-stack {
  --ld-card-stack-translate-amount-x: var(--ld-sp-6);
  --ld-card-stack-translate-amount-y: var(--ld-sp-6);
  --ld-card-stack-translate-direction-x: -1;

  display: grid;
  grid-template-areas: 'card';
  grid-template-columns: 1fr;
  padding: 0;
  position: relative;
  z-index: 0;

  ::slotted(.ld-card),
  .ld-card {
    grid-area: card;
    opacity: 0;
    transition: transform var(--ld-transition-duration-normal) ease,
      opacity var(--ld-transition-duration-normal) linear,
      visibility var(--ld-transition-duration-normal) 0s linear;
    visibility: hidden;

    &::before,
    &::after {
      opacity: 0.3;
    }
  }
  ::slotted(.ld-card:first-of-type:not(:last-of-type)),
  .ld-card:first-of-type:not(:last-of-type) {
    transform: translate(
      calc(
        var(--ld-card-stack-translate-direction-x) *
          var(--ld-card-stack-translate-amount-x)
      ),
      calc(-1 * var(--ld-card-stack-translate-amount-y))
    );
    z-index: 1;
  }
  ::slotted(.ld-card:nth-of-type(-n + 3)),
  .ld-card:nth-of-type(-n + 3) {
    opacity: 1;
    transition: transform var(--ld-transition-duration-normal) ease,
      opacity var(--ld-transition-duration-normal) linear;
    visibility: inherit;
  }
  ::slotted(.ld-card:nth-of-type(3)),
  .ld-card:nth-of-type(3) {
    transform: translate(
      calc(
        -1 * var(--ld-card-stack-translate-direction-x) * var(--ld-card-stack-translate-amount-x)
      ),
      var(--ld-card-stack-translate-amount-y)
    );
    z-index: -1;
  }
  ::slotted(.ld-card:focus-within),
  .ld-card:focus-within {
    z-index: 2;
  }

  &::after {
    content: '';
    box-shadow: var(--ld-shadow-hover);
    border-radius: var(--ld-br-l);
    position: absolute;
    inset: calc(var(--ld-card-stack-translate-amount-y) * 2)
      var(--ld-card-stack-translate-amount-x);
    z-index: -2;
    opacity: 0.8;
  }
}

:host(.ld-card-stack--rtl),
.ld-card-stack--rtl {
  --ld-card-stack-translate-direction-x: 1;
}

:host(.ld-card-stack--vertical),
.ld-card-stack--vertical {
  --ld-card-stack-translate-direction-x: 0;
}
