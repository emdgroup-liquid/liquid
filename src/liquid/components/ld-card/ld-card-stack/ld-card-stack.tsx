import { Component, h, Host, Prop } from "@stencil/core";
import { getClassNames } from "../../../utils/getClassNames";

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: "ld-card-stack",
  styleUrl: "ld-card-stack.css",
  shadow: true,
})
export class LdCardStack {
  /** The stack direction. */
  @Prop() direction?: "ltr" | "rtl" | "vertical";

  render() {
    return (
      <Host
        class={getClassNames([
          "ld-card-stack",
          `ld-card-stack--${this.direction}`,
        ])}
        role="list"
      >
        <slot></slot>
      </Host>
    );
  }
}
