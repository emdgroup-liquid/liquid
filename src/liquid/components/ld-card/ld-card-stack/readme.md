---
eleventyNavigation:
  key: Card Stack
  parent: Card
layout: layout.njk
title: Card Stack
permalink: components/ld-card/ld-card-stack/
---

<link rel="stylesheet" href="css_components/ld-card-stack.css">
<link rel="stylesheet" href="css_components/ld-card.css">

# ld-card-stack

Use the `ld-card-stack` component to display multiple [`ld-card`](components/ld-card/)s in a stack.

<ld-notice headline="Note" mode="warning">
  A maximum of three cards are displayed in the stack, even if there are more cards in it.
</ld-notice>

---

## Default

{% example %}
<ld-card-stack>
  <ld-card>Card A</ld-card>
  <ld-card>Card B</ld-card>
  <ld-card>Card C</ld-card>
  <ld-card>Card D</ld-card>
  <ld-card>Card E</ld-card>
</ld-card-stack>

<!-- CSS component -->

<ol class="ld-card-stack">
  <li class="ld-card">Card A</li>
  <li class="ld-card">Card B</li>
  <li class="ld-card">Card C</li>
  <li class="ld-card">Card D</li>
  <li class="ld-card">Card E</li>
</ol>
{% endexample %}

## Stack direction

### Left-to-right

{% example %}
<ld-card-stack direction="ltr">
  <ld-card>Card A</ld-card>
  <ld-card>Card B</ld-card>
  <ld-card>Card C</ld-card>
  <ld-card>Card D</ld-card>
  <ld-card>Card E</ld-card>
</ld-card-stack>

<!-- CSS component -->

<ol class="ld-card-stack ld-card-stack--ltr">
  <li class="ld-card">Card A</li>
  <li class="ld-card">Card B</li>
  <li class="ld-card">Card C</li>
  <li class="ld-card">Card D</li>
  <li class="ld-card">Card E</li>
</ol>
{% endexample %}

### Right-to-left

{% example %}
<ld-card-stack direction="rtl">
  <ld-card>Card A</ld-card>
  <ld-card>Card B</ld-card>
  <ld-card>Card C</ld-card>
  <ld-card>Card D</ld-card>
  <ld-card>Card E</ld-card>
</ld-card-stack>

<!-- CSS component -->

<ol class="ld-card-stack ld-card-stack--rtl">
  <li class="ld-card">Card A</li>
  <li class="ld-card">Card B</li>
  <li class="ld-card">Card C</li>
  <li class="ld-card">Card D</li>
  <li class="ld-card">Card E</li>
</ol>
{% endexample %}

### Vertical

{% example %}
<ld-card-stack direction="vertical">
  <ld-card>Card A</ld-card>
  <ld-card>Card B</ld-card>
  <ld-card>Card C</ld-card>
  <ld-card>Card D</ld-card>
  <ld-card>Card E</ld-card>
</ld-card-stack>

<!-- CSS component -->

<ol class="ld-card-stack ld-card-stack--vertical">
  <li class="ld-card">Card A</li>
  <li class="ld-card">Card B</li>
  <li class="ld-card">Card C</li>
  <li class="ld-card">Card D</li>
  <li class="ld-card">Card E</li>
</ol>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property    | Attribute   | Description                                              | Type                           | Default     |
| ----------- | ----------- | -------------------------------------------------------- | ------------------------------ | ----------- |
| `direction` | `direction` | The stack direction.                                     | `"ltr" \| "rtl" \| "vertical"` | `undefined` |
| `key`       | `key`       | for tracking the node's identity when working with lists | `string \| number`             | `undefined` |
| `ref`       | `ref`       | reference to component                                   | `any`                          | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
