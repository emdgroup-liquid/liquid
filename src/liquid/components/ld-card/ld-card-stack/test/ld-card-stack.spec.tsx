import { h } from "@stencil/core";
import { newSpecPage } from "@stencil/core/testing";
import { LdCardStack } from "../ld-card-stack";

describe("ld-card-stack", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [LdCardStack],
      template: () => (
        <ld-card-stack>
          <ld-card>Card A</ld-card>
          <ld-card>Card B</ld-card>
        </ld-card-stack>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  describe("adds classes according to props", () => {
    it("direction", async () => {
      const page = await newSpecPage({
        components: [LdCardStack],
        template: () => (
          <ld-card-stack direction="ltr">
            <ld-card>Card A</ld-card>
            <ld-card>Card B</ld-card>
          </ld-card-stack>
        ),
      });
      expect(page.root).toHaveClass("ld-card-stack--ltr");
    });
  });
});
