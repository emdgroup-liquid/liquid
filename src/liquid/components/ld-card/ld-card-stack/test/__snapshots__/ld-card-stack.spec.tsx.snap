// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-card-stack renders 1`] = `
<ld-card-stack class="ld-card-stack ld-card-stack--undefined" role="list">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ld-card>
    Card A
  </ld-card>
  <ld-card>
    Card B
  </ld-card>
</ld-card-stack>
`;
